gatsby-node.js

```diff
@@ -99,4 +106,253 @@ async function makeSynoptic(createPage, reporter, graphql) {
+      const tree = remark().parse(node.rawMarkdownBody)
+      let heading = ""
+      let content = ""
+
+      for(let i = 0; i < tree.children.length; i++){
+        if(tree.children[i].type === 'heading'){
+          if(content !== ""){
+            this.add({
+              path: node.frontmatter.path,
+              title: node.frontmatter.title,
+              heading: heading,
+              content: content
+            })
+          }
+          
+          content = ""
+          heading = tree.children[i].children[0].value
+        }else{
+          if(tree.children[i].children){
+            for(let x = 0; x < tree.children[i].children.length; x++){
+              content += getParagraph(tree.children[i].children[x])
+            }
+          }else{
+            if(tree.children[i].type === 'text'){
+              content += tree.children[i].value
+            }
+          }
+        }
+      }
+
+      this.add({
+        path: node.frontmatter.path,
+        title: node.frontmatter.title,
+        heading: heading,
+        content: content
+      })
+    })
+
+    const self = this
+
+    result_synoptic.data.allCetei.nodes.forEach(( node ) => {
+      const filePath = `${node.parent.relativeDirectory}/${node.parent.name}`
+      let dom = new JSDOM(node.prefixed, { contentType: "text/xml" })
+
+      let heading = ""
+      let content = ""
+
+      if(filePath === '/entities'){
+        if(dom.window.document.querySelector('tei-listPerson')){
+          const allElements = dom.window.document.querySelector('tei-listPerson').children
+  
+          for(let i = 0; i < allElements.length; i++){
+            if(allElements[i].tagName === 'tei-person'){
+              allElements[i].querySelectorAll('tei-note').forEach(element =>{
+                content+=element.textContent
+              })
+              self.add({
+                path: filePath,
+                title: node.parent.name,
+                heading: allElements[i].querySelector('tei-persName').textContent,
+                content: content
+              })
+              heading = ""
+              content = ""
+            }
+          }
+        }
+
+        if(dom.window.document.querySelector('tei-listPlace')){
+          const allElements = dom.window.document.querySelector('tei-listPlace').children
+  
+          for(let i = 0; i < allElements.length; i++){
+            if(allElements[i].tagName === 'tei-place'){
+              allElements[i].querySelectorAll('tei-note').forEach(element =>{
+                content+=element.textContent
+              })
+              self.add({
+                path: filePath,
+                title: node.parent.name,
+                heading: allElements[i].querySelector('tei-placeName').textContent,
+                content: content
+              })
+              heading = ""
+              content = ""
+            }
+          }
+        }
+  
+        if(dom.window.document.querySelector('tei-listOrg')){
```

Comment:
These four operations can be abstracted to a single function that takes a selector and tagname as parameters.